### PR TITLE
Add inverted index for mongodb

### PR DESCRIPTION
**Pull request by mymusise**
Signed-off-by: corcassia <mymusise1@gmail.com>
Hi, I add the Inverted Index for mongodb. What new:
- Creating the Inverted Index when training dataset
- It will just compare the similar statement when getting a response (Only `BestMatch`)

Improving performance：

| Version | Response Time(s) |
| ---------- | ------------------------ |
| Before | 3.1 |
| After | 0.1 |

It test with 28000 rows dataset on Ubuntu 14.04 + Core i5 4670 + 12G Menery.

Hope it helps.

### CODE REVIEW

**Comment by vkosuri** [comment]
It seems to me the build failed due to network issue, i have restarted I have restarted the job, hope it will pass

**Comment by corcassia** [comment]
@gunthercox  Hi, it's strange that the CI failed with `zipfile.BadZipfile: File is not a zip file` in `#2677.1`.
But I run `tox` pass all test in my python2.7 environment on Ubuntu 14.04.
Have you meet it before?

**Comment by vkosuri** [comment]
Some useful reference about invert index https://stackoverflow.com/a/8360932/358458

**Comment by vkosuri** [comment]
@corcassia Thank you for you PR, could you please clarify the different between [mongodb text indexing](https://docs.mongodb.com/v3.2/core/index-text/) and this PR invert indexing

**Comment by corcassia** [comment]
@vkosuri  Ok, my pleasure.

Before this PR, I found it will compare all samples when calling `ChatBot.get_response`. So, if we have trained a large dataset, it will spend many time to response.
`mongodb text indexing` is a `forward index`, in this change, I hope it just select the similar sample from database by `invert indexing`, rather than select all.

For example:

**Comment by vkosuri** [comment]
@corcassia Thank you, I am not sure with performance. 

To improve @gunthercox came with an idea **tag filtering** to speed up retrieval process https://github.com/gunthercox/ChatterBot/issues/925.

**Comment by corcassia** [comment]
@vkosuri  Yes, I think **tag filtering** is a good idea. 
But, to get the tag of input statement, will it iterate all sample or pass a **classify model**?

**Comment by gunthercox** [comment]
@corcassia Thank you, I'm really impressed with the performance improvements after this change. I'm going to pull these changes down locally and test out a few things. I might have further questions.

**Comment by gunthercox** [comment]
@corcassia Apologies for my delay, I will be reviewing this pull request as soon as possible.

File: chatterbot/storage/mongodb.py [hunk 1]
```diff
@@ -331,27 +381,43 @@ def remove(self, statement_text):
...
+            _statement_query = {
+                'text': {
+                    '$in': responses
+                }
+            }
+        else:
+            # Just filter the statement in need.
+            statement_ids = []
+            tokens = self.statement_segmentation(statement.text)
+            word_dicts = self.inverted.find({'word': {'$in': tokens}})
```
Comment (gunthercox) [inline chatterbot/storage/mongodb.py]: I might be misunderstanding this, so feel free to correct me if I'm wrong. Given this change, wont only statements in which the words of `statement` exist be returned?

Comment (corcassia) [reply chatterbot/storage/mongodb.py]: Yes, you're right. So this change will lead to no response if there's no statement exist.

**Comment by corcassia** [comment]
@gunthercox It's all right, and I found some problem with this change in these days.

**Comment by corcassia** [comment]
@vkosuri Hey, I was wrong. `MongoDB text indexing` is one of the practices of `invert indexing`.  After a few test, I found `MongoDB text indexing` behave better. 
I remove my `invert indexing` and change the `_statement_query`.

**Comment by zxsimple** [comment]
@mymusise seems the change hasn't checked in to the master? can I apply the path master brach for resolving the performance issue?

**Comment by mymusise** [comment]
@zxsimple Yeah, seems the master branch still not create a full-text index when setting up the bot [(chatterbot/storage/mongodb.py)](https://github.com/mymusise/ChatterBot/blob/master/chatterbot/storage/mongodb.py#L115), but you can create it manually and with a little change [like this](https://github.com/mymusise/ChatterBot/blob/master/chatterbot/storage/mongodb.py#L385).

If changing source code is a boresome work, there's another way to resolving the performance.

I think the performance with highly improve with decreasing the [`statement_list`](https://github.com/mymusise/ChatterBot/blob/master/chatterbot/logic/best_match.py#L35), cause [the comparison](https://github.com/mymusise/ChatterBot/blob/master/chatterbot/logic/best_match.py#L36) will cost a lot of CPU time.I found it may help by removing some meaningless works such as ('the', 'a', 'an', 'with', 'to') from input statement and corpus.

## Solution:
```python
import re

source_corpus = [
    "Hello",
    "Hi there!",
    "How are you doing?",
    "I'm doing great.",
    "That is good to hear",
    "Thank you.",
    "You're welcome."
]
input_statements = [
    "Such a good day!",
    "Here we are."
]

def words_filter(statements):
    # a words filter demo
    bad_words = ['the', 'a', 'an', 'is', 'are']
    rule = re.compile(" | ".join(bad_words))
    return [rule.sub(' ', statement) for statement in statements]

courpus = words_filter(source_corpus)
input_statements = words_filter(input_statements)

from chatterbot.trainers import ListTrainer
from chatterbot import ChatBot

chatbot = ChatBot("Ron Obvious")

chatbot.set_trainer(ListTrainer)
chatbot.train(courpus)


for statement in input_statements:
    response = chatbot.get_response(statement)
    print(response)
```

**Comment by zxsimple** [comment]
@mymusise ， thanks for your reply, I tried and found the bottleneck related to the number of question dataset. I have 3K+ questions, each of the comparison takes 1ms, so each response to the question takes around 3 seconds. The complexity of comparison is O(n), I'll try KNN for similarity comparison and see. 

`        for statement in statement_list:
            confidence = self.compare_statements(input_statement, statement)
            if confidence > closest_match.confidence:
                statement.confidence = confidence
                closest_match = statement`

**Pull request closed without merging**